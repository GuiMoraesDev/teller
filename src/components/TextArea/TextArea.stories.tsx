import React from 'react';

import { ComponentMeta, ComponentStory } from '@storybook/react';

import TextArea from '.';

const Storie = {
  title: 'Components/Text area',
  component: TextArea,
} as ComponentMeta<typeof TextArea>;

export default Storie;

const Template: ComponentStory<typeof TextArea> = (args) => (
  <TextArea {...args} />
);

export const Default = Template.bind({});

Default.args = {};
