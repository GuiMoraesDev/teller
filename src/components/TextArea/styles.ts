import styled, { css } from "styled-components";

import { TextAreaDefaultPropsThatMakeStyles, TextAreaProps } from ".";

export const Container = styled.div`
  position: relative;

  width: ${({ theme }) => theme.sizes.partition["x1/1"]};
`;

export const Label = styled.label``;

interface ErrorMessageProps {
  error?: string;
}

export const ErrorMessage = styled.span<ErrorMessageProps>`
  position: absolute;
  bottom: ${({ theme }) => theme.sizes.common["x0.5"]};
  right: ${({ theme }) => theme.sizes.common.x0};
  left: ${({ theme }) => theme.sizes.common.x0};

  display: flex;

  align-items: center;

  width: ${({ theme }) => theme.sizes.partition["x1/1"]};

  ${({ theme }) => theme.typography.variants.body3};

  margin: auto;
  padding: ${({ theme }) =>
    `${theme.sizes.common["x1.5"]} ${theme.sizes.common.x3}`};

  color: ${({ theme }) => theme.colors.neutrals["000"]};
  background-color: ${({ theme }) => theme.colors.alert["500"]};

  border-radius: 0;
  border-bottom-left-radius: ${({ theme }) => theme.rounded.sm};
  border-bottom-right-radius: ${({ theme }) => theme.rounded.sm};

  opacity: ${({ error }) => (error ? 1 : 0)};
  transform: translateY(100%);
  transition: opacity 0.5s ease-in-out;

  z-index: 0;

  svg {
    margin-right: 0.5rem;
  }
`;

export const ContainerTextArea = styled.div<TextAreaDefaultPropsThatMakeStyles>`
  position: relative;

  width: ${({ theme }) => theme.sizes.partition["x1/1"]};

  padding: ${({ theme }) => theme.sizes.common.x4};

  background-color: ${({ theme }) => theme.themeColors.canvas};

  border: ${({ theme }) => theme.borders.solid};
  border-radius: ${({ theme }) => theme.rounded.sm};

  transition: border-color ${({ theme }) => theme.transition.fast},
    box-shadow ${({ theme }) => theme.transition.fast};

  &:hover {
    border-color: ${({ theme }) => theme.colors.neutrals[400]};
    box-shadow: ${({ theme }) => theme.shadows.short};
  }

  ${({ theme, rounded }) => {
    return css`
      border-radius: ${theme.rounded[rounded || "none"]};
    `;
  }}

  ${({ dimension, theme }) => {
    if (dimension === "sm")
      return css`
        ${theme.typography.variants.body3};

        height: ${theme.sizes.common.x24};
      `;

    if (dimension === "md")
      return css`
        ${theme.typography.variants.body3};

        height: ${theme.sizes.common.x48};
      `;

    if (dimension === "lg")
      return css`
        ${theme.typography.variants.body2};

        height: ${theme.sizes.common.x64};
      `;
  }}

  ${({ theme, isDisabled }) => {
    if (isDisabled)
      return css`
        background-color: ${theme.colors.neutrals[200]};
        color: ${theme.colors.neutrals[500]};
        box-shadow: 0 0 0 1px ${theme.colors.neutrals[200]};

        &:hover {
          background-color: ${theme.colors.neutrals[200]};
          color: ${theme.colors.neutrals[500]};
        }
      `;
  }}

  ${({ theme, error }) => {
    if (error)
      return css`
        border-color: ${theme.colors.alert["500"]};
      `;
  }}
`;

export const TextAreaComponent = styled.textarea<TextAreaProps>`
  width: ${({ theme }) => theme.sizes.partition["x1/1"]};
  height: 85%;

  padding: 0;

  color: ${({ theme }) => theme.themeColors.text};
  background-color: ${({ theme }) => theme.themeColors.canvas};

  border: none;

  resize: none;

  ${({ dimension, theme }) => {
    if (dimension === "sm")
      return css`
        ${theme.typography.variants.body3};
      `;

    if (dimension === "md")
      return css`
        ${theme.typography.variants.body3};
      `;

    if (dimension === "lg")
      return css`
        ${theme.typography.variants.body2};
      `;
  }}
`;
