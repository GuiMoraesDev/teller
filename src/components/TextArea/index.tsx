import React, { HTMLAttributes, forwardRef } from "react";

import { Warning } from "phosphor-react";
import { DefaultTheme } from "styled-components";

import Text from "components/Text";

import * as Styles from "./styles";

export interface TextAreaDefaultPropsThatMakeStyles {
  dimension?: "sm" | "md";
  rounded?: keyof DefaultTheme["rounded"];
  isDisabled?: boolean;
  error?: string;
}

export interface TextAreaProps
  extends TextAreaDefaultPropsThatMakeStyles,
    HTMLAttributes<HTMLTextAreaElement> {
  id: string;
  label?: string;
  maxLength?: number;
  required?: boolean;
  optional?: boolean;
}

const TextArea: React.ForwardRefRenderFunction<
  HTMLTextAreaElement,
  TextAreaProps
> = (
  {
    id,
    error,
    dimension = "md",
    rounded = "sm",
    isDisabled = false,
    maxLength = 1000,
    label,
    required = false,
    optional = true,
    placeholder,
    ...props
  },
  ref
) => {
  const [charCounter, setCharCounter] = React.useState(0);

  const onContentChange = React.useCallback(
    (event: React.FormEvent<HTMLTextAreaElement>) => {
      props.onChange?.(event);

      setCharCounter(event.currentTarget.value.length);
    },
    [props]
  );

  return (
    <Styles.Container>
      {label && (
        <Styles.Label htmlFor={id}>
          <Text label={label} dimension="body1" />
        </Styles.Label>
      )}
      <Styles.ContainerTextArea
        error={error}
        dimension={dimension}
        rounded={rounded}
        isDisabled={isDisabled}
      >
        <Styles.TextAreaComponent
          id={id}
          ref={ref}
          placeholder={placeholder || "Type here..."}
          maxLength={maxLength}
          dimension={dimension}
          {...props}
          onChange={(event) => onContentChange(event)}
        />

        <Text label={`${charCounter} / ${maxLength}`} />

        <Styles.ErrorMessage error={error} data-testid="error-alert">
          <Warning />
          {error}
        </Styles.ErrorMessage>
      </Styles.ContainerTextArea>
    </Styles.Container>
  );
};

export default forwardRef(TextArea);
