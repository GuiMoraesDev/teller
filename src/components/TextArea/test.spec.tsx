import React from 'react'

import { render } from 'tests/test.utils'

import TextArea from './index'

describe('TextArea component', () => {
  it('should render correctly', () => {
    const { container } = render(<TextArea id="test-TextArea" />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render visible', () => {
    const { container } = render(<TextArea id="test-TextArea" />)

    expect(container.firstChild).toBeVisible()
  })
})
