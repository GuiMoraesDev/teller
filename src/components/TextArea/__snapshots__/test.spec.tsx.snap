// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea component should render correctly 1`] = `
.c3 {
  position: relative;
  padding: 0;
  margin: 0;
  font-size: 1.4rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  line-height: 2.0rem;
  color: #181917;
  text-transform: none;
  box-sizing: border-box;
}

.c0 {
  position: relative;
  width: 100%;
}

.c4 {
  position: absolute;
  bottom: 0.2rem;
  right: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  font-size: 1.2rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  line-height: 1.6rem;
  margin: auto;
  padding: 0.6rem 1.2rem;
  color: #FDFDFD;
  background-color: #FFD166;
  border-radius: 0;
  border-bottom-left-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
  opacity: 0;
  -webkit-transform: translateY(100%);
  -ms-transform: translateY(100%);
  transform: translateY(100%);
  -webkit-transition: opacity 0.5s ease-in-out;
  transition: opacity 0.5s ease-in-out;
  z-index: 0;
}

.c4 svg {
  margin-right: 0.5rem;
}

.c1 {
  position: relative;
  width: 100%;
  padding: 1.6rem;
  background-color: #FDFDFD;
  border: 0.1rem solid #C9CAC8;
  border-radius: 0.4rem;
  -webkit-transition: border-color 0.2s, box-shadow 0.2s;
  transition: border-color 0.2s, box-shadow 0.2s;
  border-radius: 0.4rem;
  font-size: 1.2rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  line-height: 1.6rem;
  height: 19.2rem;
}

.c1:hover {
  border-color: #949591;
  box-shadow: 1px 0px 3px 0px #E4E5E3;
}

.c2 {
  width: 100%;
  height: 85%;
  padding: 0;
  color: #181917;
  background-color: #FDFDFD;
  border: none;
  resize: none;
  font-size: 1.2rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  line-height: 1.6rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <textarea
      class="c2"
      id="test-TextArea"
      maxlength="1000"
      placeholder="Type here..."
    />
    <p
      class="c3"
    >
      0 / 1000
    </p>
    <span
      class="c4"
      data-testid="error-alert"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="128"
          x2="128"
          y1="104"
          y2="144"
        />
        <path
          d="M114.2,40l-88,152A16,16,0,0,0,40,216H216a16,16,0,0,0,13.8-24l-88-152A15.9,15.9,0,0,0,114.2,40Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <circle
          cx="128"
          cy="180"
          r="12"
        />
      </svg>
    </span>
  </div>
</div>
`;
